src/compute/src/render/mod.rs

```diff
@@ -163,13 +191,24 @@ pub fn build_compute_dataflow<A: Allocate>(
             for (source_id, (source, _monotonic)) in dataflow.source_imports.iter() {
                 // Note: For correctness, we require that sources only emit times advanced by
                 // `dataflow.as_of`. `persist_source` is documented to provide this guarantee.
-                let (ok_stream, err_stream, token) = persist_source::persist_source(
+                let (mut ok_stream, err_stream, token) = persist_source::persist_source(
                     region,
                     Arc::clone(&compute_state.persist_clients),
                     source.storage_metadata.clone(),
                     dataflow.as_of.clone().unwrap(),
                 );
 
+                // If logging is enabled, intercept frontier advancements coming from persist to track materialization lags.
+                // Note that we do this here instead of in the server.rs worker loop since we want to catch the wall-clock
+                // time of the frontier advancement for each dataflow as early as possible.
+                if let Some(logger) = compute_state.compute_logger.as_ref() {
+                    let index_ids = dataflow.index_exports.keys().cloned().collect::<Vec<_>>();
+                    let sink_ids = dataflow.sink_exports.keys().cloned().collect::<Vec<_>>();
+                    ok_stream = intercept_source_instantiation_frontiers(
+                        &ok_stream, logger, source_id, &index_ids, &sink_ids,
+                    );
```

Comment:
There is some unnecessary allocating going on here: You have two heap allocations for `index_ids` and `sink_ids`. And then inside `intercept_source_instantiation_frontiers` you allocate `index_ids` again. This can be reduced to a single allocation:

```rust
                    let index_ids = dataflow.index_exports.keys().copied();
                    let sink_ids = dataflow.sink_exports.keys().copied();
                    let export_ids = index_ids.chain(sink_ids).collect();
                    ok_stream = intercept_source_instantiation_frontiers(
                        &ok_stream, logger.clone(), source_id, export_ids,
                    );
```

The first three lines can also be replaced by a method call:

```rust
                    let export_ids = dataflow.export_ids().collect();
                    ok_stream = intercept_source_instantiation_frontiers(
                        &ok_stream, logger.clone(), source_id, export_ids,
                    );
```

Note how this also clones the logger in the caller, instead of passing it by reference and cloning inside the callee. It's generally good style for functions that need owned values to take owned values as arguments, so the cost of cloning is not hidden from the caller. (Of course in this case, cloning the logger doesn't cost much.) See [the Rust API Guidlines](https://rust-lang.github.io/api-guidelines/flexibility.html#caller-decides-where-to-copy-and-place-data-c-caller-control).